cmd2/cmd2.py

```diff
@@ -528,43 +531,43 @@ def ppaged(self, msg: str, end: str = '\n', chop: bool = False) -> None:
-                if functional_terminal and not self._redirecting and not self.in_pyscript() and not self.in_script():
-                    if ansi.allow_ansi.lower() == ansi.ANSI_NEVER.lower():
-                        msg_str = ansi.strip_ansi(msg_str)
-
-                    pager = self.pager
-                    if chop:
-                        pager = self.pager_chop
-
-                    # Prevent KeyboardInterrupts while in the pager. The pager application will
-                    # still receive the SIGINT since it is in the same process group as us.
-                    with self.sigint_protection:
-                        pipe_proc = subprocess.Popen(pager, shell=True, stdin=subprocess.PIPE)
-                        pipe_proc.communicate(msg_str.encode('utf-8', 'replace'))
-                else:
-                    self.poutput(msg_str, end='')
-            except BrokenPipeError:
-                # This occurs if a command's output is being piped to another process and that process closes before the
-                # command is finished. If you would like your application to print a warning message, then set the
-                # broken_pipe_warning attribute to the message you want printed.`
-                if self.broken_pipe_warning:
-                    sys.stderr.write(self.broken_pipe_warning)
+        if not msg:
+            return
+
+        try:
+            msg_str = '{}{}'.format(msg, end)
+
+            # Attempt to detect if we are not running within a fully functional terminal.
+            # Don't try to use the pager when being run by a continuous integration system like Jenkins + pexpect.
+            functional_terminal = False
+
+            if self.stdin.isatty() and self.stdout.isatty():
+                if sys.platform.startswith('win') or os.environ.get('TERM') is not None:
+                    functional_terminal = True
+
+            # Don't attempt to use a pager that can block if redirecting or running a script (either text or Python)
+            # Also only attempt to use a pager if actually running in a real fully functional terminal
+            if functional_terminal and not self._redirecting and not self.in_pyscript() and not self.in_script():
+                if ansi.allow_ansi.lower() == ansi.ANSI_NEVER.lower():
+                    msg_str = ansi.strip_ansi(msg_str)
+
+                pager = self.pager
+                if chop:
+                    pager = self.pager_chop
+
+                # Prevent KeyboardInterrupts while in the pager. The pager application will
+                # still receive the SIGINT since it is in the same process group as us.
+                with self.sigint_protection:
+                    pipe_proc = subprocess.Popen(pager, shell=True, stdin=subprocess.PIPE)
+                    pipe_proc.communicate(msg_str.encode('utf-8', 'replace'))
+            else:
+                self.poutput(msg_str, end='')
```

Comment:
Should this be `end=end`?

Reply:
Good catch. I wonder how long it's been like that.